>[Functional stateless tests (ubsan) — Timeout, fail: 0, passed: 0](https://clickhouse-test-reports.s3.yandex.net/30826/c6dd3656de50aa84c0d91c91bc7322b5d8ed2c46/functional_stateless_tests_(ubsan)/test_run.txt.out.log)

```
2021-10-30 00:12:21 clickhouse-test: error: unrecognized arguments: --check-zookeeper-session
```

>[CIGithubActions / FunctionalStatelessTestDebug (pull_request) Failing after 59m — FunctionalStatelessTestDebug](https://github.com/ClickHouse/ClickHouse/runs/4048549233?check_suite_focus=true)

```
Traceback (most recent call last):
  File "functional_test_check.py", line 281, in <module>
    state, description, test_results, additional_logs = process_results(result_path, server_log_path)
  File "functional_test_check.py", line 179, in process_results
    with open(status_path, 'r', encoding='utf-8') as status_file:
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/actions-runner/_work/_temp/stateless_debug/result_path/check_status.tsv'
```

--- a/src/Storages/StorageReplicatedMergeTree.cpp
+++ b/src/Storages/StorageReplicatedMergeTree.cpp
@@ -849,12 +849,47 @@ void StorageReplicatedMergeTree::dropReplica(zkutil::ZooKeeperPtr zookeeper, con
         throw Exception("Table was not dropped because ZooKeeper session has expired.", ErrorCodes::TABLE_WAS_NOT_DROPPED);
 
     auto remote_replica_path = zookeeper_path + "/replicas/" + replica;
+
     LOG_INFO(logger, "Removing replica {}, marking it as lost", remote_replica_path);
     /// Mark itself lost before removing, because the following recursive removal may fail
     /// and partially dropped replica may be considered as alive one (until someone will mark it lost)
     zookeeper->trySet(remote_replica_path + "/is_lost", "1");
+
+    /// NOTE: we should check for remote_replica_path existence,
+    /// since otherwise DROP REPLICA will fail if the replica had been already removed.
+    if (!zookeeper->exists(remote_replica_path))
+    {
+        LOG_INFO(logger, "Removing replica {} does not exist", remote_replica_path);
+        return;
+    }
+
+    /// Analog of removeRecursive(remote_replica_path)
+    /// but it removes "metadata" firstly.
+    ///
+    /// This will allow to mark table as readonly
+    /// and skip any checks of parts between on-disk and in the zookeeper.
+    ///
+    /// Without this removeRecursive() may remove "parts" first
+    /// and on DETACH/ATTACH (or server restart) it will trigger the following error:
+    ///
+    ///       "The local set of parts of table X doesn't look like the set of parts in ZooKeeper"
+    ///
+    {
+        Strings children = zookeeper->getChildren(remote_replica_path);
+
+        if (std::find(children.begin(), children.end(), "metadata") != children.end())
+            zookeeper->remove(fs::path(remote_replica_path) / "metadata");
+
+        for (const auto & child : children)
+        {
+            if (child != "metadata")
+                zookeeper->removeRecursive(fs::path(remote_replica_path) / child);
+        }
+
+        zookeeper->remove(remote_replica_path);
+    }
+
     /// It may left some garbage if replica_path subtree are concurrently modified
-    zookeeper->tryRemoveRecursive(remote_replica_path);
     if (zookeeper->exists(remote_replica_path))
         LOG_ERROR(logger, "Replica was not completely removed from ZooKeeper, {} still exists and may contain some garbage.", remote_replica_path);
 

--- a/src/Storages/StorageReplicatedMergeTree.cpp
+++ b/src/Storages/StorageReplicatedMergeTree.cpp
@@ -852,7 +852,7 @@ void StorageReplicatedMergeTree::dropReplica(zkutil::ZooKeeperPtr zookeeper, con
     LOG_INFO(logger, "Removing replica {}, marking it as lost", remote_replica_path);
     /// Mark itself lost before removing, because the following recursive removal may fail
     /// and partially dropped replica may be considered as alive one (until someone will mark it lost)
-    zookeeper->trySet(zookeeper_path + "/replicas/" + replica + "/is_lost", "1");
+    zookeeper->trySet(remote_replica_path + "/is_lost", "1");
     /// It may left some garbage if replica_path subtree are concurrently modified
     zookeeper->tryRemoveRecursive(remote_replica_path);
     if (zookeeper->exists(remote_replica_path))

